client/pages/tables.jsx

```diff
@@ -0,0 +1,131 @@
+import React from 'react';
+import AppContext from '../lib/app-context';
+// import Redirect from '../components/redirect';
+import Navbar from '../components/navbar';
+import ListTables from '../components/listTables';
+
+export default class Tables extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      tablesFetch: [],
+      isLoading: true,
+      isFreeModalSelected: false,
+      isAssignModalSelected: false,
+      tableSelected: null
+    };
+    this.freeTable = this.freeTable.bind(this);
+    this.assignTable = this.assignTable.bind(this);
+    this.noBtnClick = this.noBtnClick.bind(this);
+    this.yesBtnClick = this.yesBtnClick.bind(this);
+
+  }
+
+  componentDidMount() {
+    fetch('/api/get/tables')
+      .then(res => res.json())
+      .then(result => {
+        this.setState({ tablesFetch: result, isLoading: false });
+      });
+  }
+
+  freeTable(event) {
+    this.setState({ isFreeModalSelected: true, tableSelected: event.target.id });
+  }
+
+  assignTable(event) {
+    this.setState({ isAssignModalSelected: true, tableSelected: event.target.id });
+  }
+
+  noBtnClick(event) {
+    this.setState({ [event.target.name]: false, tableSelected: null });
+  }
+
+  yesBtnClick(event) {
+    if (event.target.name === 'isFreeModalSelected') {
+      const req = {
+        method: 'PUT',
+        headers: {
+          'Content-Type': 'application/json'
+        },
+        body: null
+      };
+      fetch(`/api/put/table/${this.state.tableSelected}`, req)
+        .then(res => res.json())
+        .then(result => {
+        });
+      for (let i = 0; i < this.state.tablesFetch.length; i++) {
+        if (this.state.tablesFetch[i].tableNumber === parseInt(this.state.tableSelected)) {
+          const newdata = this.state.tablesFetch.slice(i, i + 1);
+          newdata[0].userId = null;
+          const newstate = this.state.tablesFetch.slice(0, i).concat(newdata[0]).concat(this.state.tablesFetch.slice(i + 1));
+          this.setState({ tablesFetch: newstate, isFreeModalSelected: false, tableSelected: null });
+        }
+      }
```

Comment:
This loop can be changed to two parts:

- calculate the new array of tables using `.map()`
- set the state

If you move this into the `.then()` callback, you can use the response body from the server in place of the old table.

```js
const newTables = this.state.tablesFetch.map(oldTable => {
  if (oldTable.tableId === this.state.tableSelected) {
    return newTable
  }
  return oldTable
})
```